### PR TITLE
fixes

### PR DESCRIPTION
**Pull request by jckhang**
see https://github.com/Accern/xyme-backend/pull/35

### CODE REVIEW

File: accern_xyme/accern_xyme.py [hunk 1]
```diff
@@ -1393,7 +1396,7 @@ def as_str(self) -> str:
     def download_zip(self, save_path: str) -> None:
```
Comment (JosuaKrause) [inline accern_xyme/accern_xyme.py:1396]: don't use save path or make it optional

File: accern_xyme/accern_xyme.py [hunk 2]
```diff
@@ -1393,7 +1396,7 @@ def as_str(self) -> str:
...
         if self.is_full():
             raise ValueError(f"URI must not be full: {self}")
-        cur_res, ctype = self._client._raw_request_bytes(
+        cur_res, _ = self._client._raw_request_bytes(
             METHOD_GET, "/download_zip", {
                 "blob": self._uri,
                 "pipeline": self.get_pipeline().get_id(),
```
Comment (JosuaKrause) [inline accern_xyme/accern_xyme.py:1402]: allow upload of bytesio as zip

File: accern_xyme/util.py [hunk 3]
```diff
@@ -177,7 +177,10 @@ def get_file_hash(buff: IO[bytes]) -> str:
 
 def interpret_ctype(data: IO[bytes], ctype: str) -> ByteResponse:
     if ctype == "application/json":
-        return json.load(data)
+        res = json.load(data)
+        if "errMessage" in res and res["errMessage"]:
```
Comment (JosuaKrause) [inline accern_xyme/util.py:181]: hmm, that's not good -- should be a different ctype then -- how do we know the actual output doesn't contain a "errMessage" field?

Comment (JosuaKrause) [reply accern_xyme/util.py:181]: do application/json+error

File: accern_xyme/accern_xyme.py [hunk 4]
```diff
@@ -807,12 +812,25 @@ def dynamic_obj(self, input_obj: Any) -> ByteResponse:
 
     def dynamic_async(
             self, input_data: List[BytesIO]) -> List['ComputationHandle']:
+        self._dynamic_error = None
```
Comment (JosuaKrause) [inline accern_xyme/accern_xyme.py:815]: why do we store it quasi globally?

Comment (JosuaKrause) [reply accern_xyme/accern_xyme.py:815]: especially since you are unconditionally clearing it

File: accern_xyme/accern_xyme.py [hunk 5]
```diff
@@ -833,12 +851,34 @@ def get_dynamic_result(self, data_id: str) -> ByteResponse:
...
+            data_ids: List['ComputationHandle']) -> Dict[
+                'ComputationHandle', QueueStatus]:
+        res = cast(DynamicStatusResponse, self._client._request_json(
+            METHOD_POST, "/dynamic_status", {
+                "data_ids": [data_id.get_id() for data_id in data_ids],
+                "pipeline": self._pipe_id,
+            }, capture_err=True))
+        status = res["status"]
+        return {
+            ComputationHandle(
```
Comment (JosuaKrause) [inline accern_xyme/accern_xyme.py:869]: can you not reuse the input computation handle?

File: accern_xyme/accern_xyme.py [hunk 6]
```diff
@@ -1164,8 +1213,13 @@ def read_blob(self, key: str, chunk: int) -> 'BlobHandle':
...
             pipeline=self.get_pipeline())
 
-    def read(self, key: str, chunk: int) -> Optional[ByteResponse]:
-        return self.read_blob(key, chunk).get_content()
+    def read(
+            self,
+            key: str,
+            chunk: int,
+            force_refresh: bool = False) -> Optional[ByteResponse]:
+        content = self.read_blob(key, chunk, force_refresh).get_content()
```
Comment (JosuaKrause) [inline accern_xyme/accern_xyme.py:1221]: hmm?

File: accern_xyme/accern_xyme.py [hunk 7]
```diff
@@ -1641,6 +1724,16 @@ def __str__(self) -> str:
     def __repr__(self) -> str:
         return f"{self.__class__.__name__}[{self.__str__()}]"
 
+    def __hash__(self) -> int:
+        return hash(self.get_id())
+
+    def __eq__(self, other: object) -> bool:
+        return self.__str__() == other.__str__()
```
Comment (JosuaKrause) [inline accern_xyme/accern_xyme.py:1731]: nope use id

File: accern_xyme/util.py [hunk 8]
```diff
@@ -220,24 +226,130 @@ def async_compute(
...
+                    do_wait = False
+                    try:
+                        t_ix = ids.pop(t_id)
+                        res[t_ix] = get(t_id)
+                    except NotFoundError:
+                        pass
+                    except ServerSideError as e:
+                        if exc[0] is None:
+                            exc[0] = e
+                    except KeyError as e:
```
Comment (JosuaKrause) [inline accern_xyme/util.py:310]: no e

File: accern_xyme/util.py [hunk 9]
```diff
@@ -220,24 +226,130 @@ def async_compute(
...
+                ])
+                for (t_id, t_status) in status.items():
+                    # print("t_id, t_status", t_id, t_status, t_id in ids)
+                    if t_status in ("waiting", "running"):
+                        continue
+                    do_wait = False
+                    try:
+                        t_ix = ids.pop(t_id)
+                        res[t_ix] = get(t_id)
+                    except NotFoundError:
```
Comment (JosuaKrause) [inline accern_xyme/util.py:305]: I thought we are doing KeyError?

File: accern_xyme/util.py [hunk 10]
```diff
@@ -220,24 +226,130 @@ def async_compute(
...
+class ServerSideError(Exception):
+    def __init__(self, message: str) -> None:
+        self._message = message
+        super().__init__(self._message)
+
+    def __str__(self) -> str:
+        return f"Error from xyme backend: \n{self._message}"
+
+
+class NotFoundError(Exception):
```
Comment (JosuaKrause) [inline accern_xyme/util.py:353]: so the plan was to actually raise an exception on a 404 (in any request method) and then convert that exception to a keyerror for the dynamic getter

File: accern_xyme/util.py [hunk 11]
```diff
@@ -220,24 +226,133 @@ def async_compute(
...
+    end_produce: List[bool] = [False]
+    exc: List[Optional[Exception]] = [None]
+    cond = threading.Condition()
+    ids: Dict[RT, int] = {}
+    res: Dict[int, ByteResponse] = {}
+
+    def get_max_count(remote_queue: QueueStatsResponse) -> int:
+        return max(
+            remote_queue["total"] - remote_queue["active"],
+            remote_queue["data"] - remote_queue["active"])
```
Comment (JosuaKrause) [inline accern_xyme/util.py:247]: this is wrong

Comment (JosuaKrause) [reply accern_xyme/util.py:247]: why data-active?

Comment (JosuaKrause) [reply accern_xyme/util.py:247]:
```
def get_data_count(self) -> int:
        return len(self.obj_keys(self._data_lookup))

    def get_total_count(self) -> int:
        return self.get_data_count() + sum((el[1] for el in self.get_queues()))
```

Comment (JosuaKrause) [reply accern_xyme/util.py:247]: total includes the queues that haven't been processed yet

File: accern_xyme/util.py [hunk 12]
```diff
@@ -220,24 +226,137 @@ def async_compute(
...
+            do_wait = False
+            while ids:
+                if do_wait:
+                    time.sleep(1)
+                do_wait = True
+                check_ids = [
+                    v[0]
+                    for v in sorted(ids.items(), key=lambda v: v[1]) \
+                        [th_id:th_id + 10 * num_threads:num_threads]]
+                status = get_status(check_ids)
```
Comment (JosuaKrause) [inline accern_xyme/util.py:300]: test against empty check_ids

File: accern_xyme/util.py [hunk 13]
```diff
@@ -220,24 +226,136 @@ def async_compute(
...
+        while not done[0]:
+            with cond:
+                cond.wait_for(
+                    lambda: exc[0] is not None or done[0] or len(ids) > 0)
+            do_wait = False
+            while ids:
+                if do_wait:
+                    time.sleep(1)
+                do_wait = True
+                val = (th_id, th_id + 10 * num_threads, num_threads)
```
Comment (JosuaKrause) [inline accern_xyme/util.py:294]: range(...)

File: accern_xyme/util.py [hunk 14]
```diff
@@ -220,24 +226,136 @@ def async_compute(
...
-        yield get(cur_id)
+    assert num_threads > 0
+    done: List[bool] = [False]
+    end_produce: List[bool] = [False]
+    exc: List[Optional[Exception]] = [None]
+    cond = threading.Condition()
+    ids: Dict[RT, int] = {}
+    res: Dict[int, ByteResponse] = {}
+
+    def get_max_count(remote_queue: QueueStatsResponse) -> int:
```
Comment (JosuaKrause) [inline accern_xyme/util.py:244]: call it waiting_count

File: accern_xyme/util.py (hunk 13 above)
Comment (JosuaKrause) [reply accern_xyme/util.py:294]: slice not range

Comment (JosuaKrause) [reply accern_xyme/util.py:294]: `sub = slice(th_id, th_id + 10 * num_threads, num_threads)`

File: accern_xyme/util.py [hunk 15]
```diff
@@ -220,24 +226,136 @@ def async_compute(
...
+                cond.wait_for(
+                    lambda: exc[0] is not None or done[0] or len(ids) > 0)
+            do_wait = False
+            while ids:
+                if do_wait:
+                    time.sleep(1)
+                do_wait = True
+                val = (th_id, th_id + 10 * num_threads, num_threads)
+                sorted_ids = sorted(ids.items(), key=lambda v: v[1])
+                check_ids = [v[0] for v in sorted_ids[val[0]:val[1]:val[2]]]
```
Comment (JosuaKrause) [inline accern_xyme/util.py:296]: `sorted_ids[sub]`

File: accern_xyme/util.py (hunk 14 above)
Comment (JosuaKrause) [reply accern_xyme/util.py:244]: get_waiting_count

File: accern_xyme/accern_xyme.py [hunk 16]
```diff
@@ -1620,18 +1687,35 @@ def append_jsons(self, jsons: List[Any]) -> 'JSONBlobHandle':
...
+        self._set_dyn_error = set_dyn_error
 
     def has_fetched(self) -> bool:
         return self._value is not None
 
     def get(self) -> ByteResponse:
-        if self._value is None:
-            self._value = self._pipeline.get_dynamic_result(self._data_id)
-        return self._value
+        maybe_error = self._get_dyn_error()
```
Comment (JosuaKrause) [inline accern_xyme/accern_xyme.py:1706]: if it has a valid value we need to return the valid value -- also, we need to call get_dynamic_result to clear it from redis. it's just that when reporting the error we swap it out with dyn_error

File: accern_xyme/accern_xyme.py [hunk 17]
```diff
@@ -1620,18 +1687,35 @@ def append_jsons(self, jsons: List[Any]) -> 'JSONBlobHandle':
...
+        if maybe_error is not None:
+            raise ServerSideError(maybe_error)
+        try:
+            if self._value is None:
+                self._value = self._pipeline.get_dynamic_result(self._data_id)
+            return self._value
+        except ServerSideError as e:
+            self._set_dyn_error(str(e))
```
Comment (JosuaKrause) [inline accern_xyme/accern_xyme.py:1714]: only set if not set already

File: accern_xyme/accern_xyme.py [hunk 18]
```diff
@@ -826,48 +840,89 @@ def dynamic_async_obj(
...
+        from requests.exceptions import HTTPError
+        try:
+            cur_res, ctype = self._client.request_bytes(
+                METHOD_GET, "/dynamic_result", {
+                    "pipeline": self._pipe_id,
+                    "id": data_id,
+                })
+        except HTTPError as e:
+            if e.response.status_code == 404:
+                raise KeyError(f"data_id {data_id} does not exist") from e
```
Comment (JosuaKrause) [inline accern_xyme/accern_xyme.py:852]: reraise for anything else

File: accern_xyme/accern_xyme.py [hunk 19]
```diff
@@ -826,48 +840,89 @@ def dynamic_async_obj(
         ])
 
     def get_dynamic_result(self, data_id: str) -> ByteResponse:
-        cur_res, ctype = self._client.request_bytes(
-            METHOD_GET, "/dynamic_result", {
-                "pipeline": self._pipe_id,
-                "id": data_id,
-            })
+        from requests.exceptions import HTTPError
```
Comment (JosuaKrause) [inline accern_xyme/accern_xyme.py:843]: import globally

File: accern_xyme/accern_xyme.py [hunk 20]
```diff
@@ -826,48 +840,89 @@ def dynamic_async_obj(
...
+        try:
+            cur_res, ctype = self._client.request_bytes(
+                METHOD_GET, "/dynamic_result", {
+                    "pipeline": self._pipe_id,
+                    "id": data_id,
+                })
+        except HTTPError as e:
```
Comment (JosuaKrause) [inline accern_xyme/accern_xyme.py:850]: handle Server error, no?

File: accern_xyme/util.py [hunk 21]
```diff
@@ -178,6 +179,9 @@ def get_file_hash(buff: IO[bytes]) -> str:
 def interpret_ctype(data: IO[bytes], ctype: str) -> ByteResponse:
     if ctype == "application/json":
         return json.load(data)
+    if ctype == "application/json-error":
```
Comment (JosuaKrause) [inline accern_xyme/util.py:182]: json+error

File: accern_xyme/accern_xyme.py (hunk 20 above)
Comment (jckhang) [reply accern_xyme/accern_xyme.py:850]: ServerSideError? that's in interpret_ctype following

File: accern_xyme/util.py [hunk 22]
```diff
@@ -220,24 +224,129 @@ def async_compute(
...
-        while count < block_size and ids:
-            yield get(ids.popleft())
-            count += 1
-        cur_size = block_size
-    for cur_id in ids:
-        yield get(cur_id)
+    assert num_threads > 0
+    done: List[bool] = [False]
+    end_produce: List[bool] = [False]
+    exc: List[Optional[Exception]] = [None]
```
Comment (JosuaKrause) [inline accern_xyme/util.py:237]: BaseException

File: accern_xyme/accern_xyme.py [hunk 23]
```diff
@@ -327,16 +332,14 @@ def _fallible_raw_request_bytes(
...
             if req.status_code == 200:
                 return BytesIO(req.content), req.headers["content-type"]
-            raise ValueError(
-                f"error {req.status_code} in worker request:\n{req.text}")
+            req.raise_for_status()
         if method == METHOD_POST:
             req = requests.post(url, json=args)
             if req.status_code == 403:
                 raise AccessDenied(req.text)
             if req.status_code == 200:
```
Comment (jckhang) [inline accern_xyme/accern_xyme.py:340]: return bytsio after raise_for_status()

Comment (jckhang) [reply accern_xyme/accern_xyme.py:340]: and also do everywhere with requests

File: accern_xyme/accern_xyme.py [hunk 24]
```diff
@@ -826,48 +841,89 @@ def dynamic_async_obj(
...
+            raise e
         return interpret_ctype(cur_res, ctype)
 
+    def check_queue_stats(self) -> QueueStatsResponse:
+        return cast(QueueStatsResponse, self._client._request_json(
+            METHOD_GET, "/queue_stats", {}, capture_err=True))
+
+    def get_dynamic_status(
+            self,
+            data_ids: List['ComputationHandle']) -> Dict[
```
Comment (jckhang) [inline accern_xyme/accern_xyme.py:862]: ,
) -> Dict[...]

File: accern_xyme/util.py [hunk 25]
```diff
@@ -220,24 +224,128 @@ def async_compute(
...
+
+    def get_waiting_count(remote_queue: QueueStatsResponse) -> int:
+        return remote_queue["total"] - remote_queue["active"]
+
+    def can_push_more() -> bool:
+        if exc[0] is not None:
+            return True
+        if len(ids) < max_buff:
+            return True
+        max_count = get_waiting_count(check_queue())
```
Comment (jckhang) [inline accern_xyme/util.py:250]: rename

File: accern_xyme/util.py [hunk 26]
```diff
@@ -220,24 +224,128 @@ def async_compute(
...
+        try:
+            pos = 0
+            while pos < len(arr):
+                with cond:
+                    cond.wait_for(can_push_more)
+                if exc[0] is not None:
+                    break
+                start_pos = pos
+                remote_queue = check_queue()
+                max_count = get_waiting_count(remote_queue)
```
Comment (jckhang) [inline accern_xyme/util.py:263]: change max_count

File: accern_xyme/util.py [hunk 27]
```diff
@@ -220,24 +224,128 @@ def async_compute(
...
+                    do_wait = False
+                    try:
+                        t_ix = ids.pop(t_id)
+                        res[t_ix] = get(t_id)
+                    except KeyError:
+                        pass
+                    except ServerSideError as e:
+                        if exc[0] is None:
+                            exc[0] = e
+                with cond:
```
Comment (jckhang) [inline accern_xyme/util.py:309]: if do_wait:
  time.sleep(1)
else:
   notify()

**Review by JosuaKrause — APPROVED** [review]
lgtm

**Review by JosuaKrause — DISMISSED** [review]
lgtm

**Review by JosuaKrause — APPROVED** [review]
lgtm

**Pull request merged**